pkg/ha/stage.go

```diff
@@ -0,0 +1,364 @@
+		return stage, err
+	}
+	stage.Source = ks[0]
+	return stage, nil
+}
+
+// subTaskStageFromKey constructs an incomplete subtask stage from an etcd key.
+func subTaskStageFromKey(key string) (Stage, error) {
+	var stage Stage
+	ks, err := common.StageSubTaskKeyAdapter.Decode(key)
+	if err != nil {
+		return stage, err
+	}
+	stage.Source = ks[0]
+	stage.Task = ks[1]
+	return stage, nil
+}
+
+// watchStage watches PUT & DELETE operations for the stage.
+func watchStage(ctx context.Context, watchCh clientv3.WatchChan,
+	stageFromKey func(key string) (Stage, error), outCh chan<- Stage, errCh chan<- error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case resp := <-watchCh:
+			if resp.Canceled {
+				// TODO(csuzhangxc): do retry here.
+				if resp.Err() != nil {
+					select {
+					case errCh <- resp.Err():
+					case <-ctx.Done():
+					}
+				}
+				return
+			}
+
+			for _, ev := range resp.Events {
+				var (
+					stage Stage
+					err   error
+				)
+				switch ev.Type {
+				case mvccpb.PUT:
+					stage, err = stageFromJSON(string(ev.Kv.Value))
+				case mvccpb.DELETE:
+					stage, err = stageFromKey(string(ev.Kv.Key))
+					stage.IsDeleted = true
+				default:
+					// this should not happen.
+					log.L().Error("unsupported etcd event type", zap.Reflect("kv", ev.Kv), zap.Reflect("type", ev.Type))
+					continue
+				}
+				stage.Revision = ev.Kv.ModRevision
+
+				if err != nil {
+					select {
+					case errCh <- err:
+					case <-ctx.Done():
+						return
+					}
+				} else {
+					select {
+					case outCh <- stage:
+					case <-ctx.Done():
+						return
+					}
+				}
+			}
+		}
+	}
+}
+
+// putRelayStageOp returns a list of PUT etcd operation for the relay stage.
+// k/v: sourceID -> the running stage of the relay.
+func putRelayStageOp(stages ...Stage) ([]clientv3.Op, error) {
+	ops := make([]clientv3.Op, 0, len(stages))
+	for _, stage := range stages {
+		value, err := stage.toJSON()
+		if err != nil {
+			return ops, err
+		}
+		key := common.StageRelayKeyAdapter.Encode(stage.Source)
+		ops = append(ops, clientv3.OpPut(key, value))
+	}
+	return ops, nil
+}
+
+// putSubTaskStageOp returns a list of PUT etcd operations for the subtask stage.
+// k/v: sourceID, task -> the running stage of the subtask.
+func putSubTaskStageOp(stages ...Stage) ([]clientv3.Op, error) {
+	ops := make([]clientv3.Op, 0, len(stages))
+	for _, stage := range stages {
+		value, err := stage.toJSON()
+		if err != nil {
+			return ops, err
+		}
+		key := common.StageSubTaskKeyAdapter.Encode(stage.Source, stage.Task)
+		ops = append(ops, clientv3.OpPut(key, value))
+	}
+	return ops, nil
+}
+
+// deleteRelayStageOp returns a DELETE etcd operation for the relay stage.
+func deleteRelayStageOp(source string) clientv3.Op {
+	return clientv3.OpDelete(common.StageRelayKeyAdapter.Encode(source))
+}
+
+// deleteSubTaskStageOp returns a list of DELETE etcd operation for the subtask stage.
+func deleteSubTaskStageOp(stages ...Stage) []clientv3.Op {
```

Comment:
these subtasks have same task name or same source?

Reply:
may have the same task name, but often have different source IDs (like stoping a task with multi subtasks with multi sources).